packages/pigeon/lib/java_generator.dart

```diff
@@ -458,14 +478,66 @@ void generateJava(JavaOptions options, Root root, StringSink sink) {
         for (final NamedType field in klass.fields) {
           final HostDatatype hostDatatype = getHostDatatype(field, root.classes,
               root.enums, (NamedType x) => _javaTypeForBuiltinDartType(x.type));
-          indent.writeln('private ${hostDatatype.datatype} ${field.name};');
+          final String nullability =
+              field.type.isNullable ? '@Nullable' : '@NonNull';
           indent.writeln(
-              'public ${hostDatatype.datatype} ${_makeGetter(field)}() { return ${field.name}; }');
+              'private $nullability ${hostDatatype.datatype} ${field.name};');
           indent.writeln(
-              'public void ${_makeSetter(field)}(${hostDatatype.datatype} setterArg) { this.${field.name} = setterArg; }');
+              'public $nullability ${hostDatatype.datatype} ${_makeGetter(field)}() { return ${field.name}; }');
+          indent.writeScoped(
+              'public void ${_makeSetter(field)}($nullability ${hostDatatype.datatype} setterArg) {',
+              '}', () {
+            if (!field.type.isNullable) {
+              indent.writeScoped('if (setterArg == null) {', '}', () {
+                indent.writeln(
+                    'throw new IllegalStateException("Nonnull field \\"${field.name}\\" is null.");');
+              });
+            }
+            indent.writeln('this.${field.name} = setterArg;');
+          });
           indent.addln('');
         }
-        indent.write('Map<String, Object> toMap() ');
+
+        if (klass.fields
```

Comment:
Same thing here; the constructor/builder seems like a great thing to extract from this large-and-growing method.